### PR TITLE
Usage Reporting

### LINKED ISSUE #49: Report resource usage by step and overall
Our job runner should be able to report how long each step ran for and what resources it used. This will be critical for accounting but could also be used for real-time progress reporting.

@johnbradley and i discussed recently how to accomplish this, and I suggested it would be best addressed by calrissian

Some thoughts:

- My initial idea was to address progress reporting and resource usage in the same mechanism - by watching for jobs issued that match a label or annotation.
- [lando watcher](https://github.com/Duke-GCB/lando/blob/master/lando/k8s/watcher.py) watches k8s events and translates them into rabbitmq messages. This would be useful for reporting real-time updates back to bespin-api through the existing queue.
- Using lando watcher for that purpose is a stretch because lando and bespin aren't concerned with the steps of a CWL workflow or number of scattered tasks.
- Calrissian (as the CWL engine) obviously is concerned with these progress details, so it makes sense to digest and broadcast this information from calrissian
- Calrissian can easily track and report these times in a single JSON file at the end of the execution. This is a good feature for all users, not just bespin/lando integration
- Reporting real-time progress is trickier, as this would likely need to move through the message queue.

### PR DESCRIPTION
**Pull request by dleehr**
Enables reporting of memory, cpu, and runtime for a workflow

- Adds `--usage-report <report.json>` optional argument to write out a JSON report at end of workflow
- Report includes workflow summary and step-specific metrics
- Step-specific metrics show individual step CPU/memory usage as well as duration
- Workflow summary computes maximum parallel resource usage and total CPU/memory hours consumed
- Does not yet include any workflow-sourced metadata (like the CWL step name)

Fixes #49

### CODE REVIEW

File: tests/test_k8s.py [hunk 1]
```diff
@@ -90,7 +93,8 @@ def test_wait_skips_pod_when_status_is_none(self, mock_watch, mock_get_namespace
 
     @patch('calrissian.k8s.watch')
     def test_wait_skips_pod_when_state_is_running(self, mock_watch, mock_get_namespace, mock_client):
-        mock_pod = Mock(status=Mock(container_statuses=[Mock(state=Mock(running=Mock(), terminated=None, waiting=None))]))
+        mock_pod = Mock(
+            status=Mock(container_statuses=[Mock(state=Mock(running=Mock(), terminated=None, waiting=None))]))
```
Comment (johnbradley) [inline tests/test_k8s.py]: You might be able to simplify creating `mock_pod` into
```
mock_pod = create_autospec(V1Pod)
```
After importing `V1Pod`: `from kubernetes.client.models import V1Pod`
And  create_autospec from mock: `from unittest.mock import create_autospec`

File: calrissian/k8s.py [hunk 2]
```diff
@@ -173,6 +208,7 @@ def get_current_pod(self):
         return self.get_pod_for_name(pod_name)
 
 
+
```
Comment (johnbradley) [inline calrissian/k8s.py]: Extra blank line.

File: calrissian/report.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    url = 'https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-cpu'
+    suffixes = {
+        'm': 0.001
+    }
+
+
+class TimedResourceReport(TimedReport):
+    """
+    Adds CPU and memory usage to TimedReport, in order to calculate resource
+    usage over the duration of the timed report
```
Comment (johnbradley) [inline calrissian/report.py]: I wasn't sure if the CPU and memory values here were actual or requested.
Perhaps a comment to clarify that.

**Comment by dleehr** [comment]
@johnbradley I fixed #52 and added those values to the report too

**Pull request merged**